File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.cpp
struct Node {
       Node* list[26];
       bool End = false;
       void assignNode(char c , Node* node)
       {
        list[c-'a']=node;
       }
       Node* returnNode(char c)
       {
         return list[c-'a'];
       }
       bool isEnd(){
         return End;
       }
       void setEnd(){
         End = true;
       }
    };
class Trie {
      Node* root;
public:
        Trie(){
            root = new Node();
        }
        
    
    void insert(string word) {
        Node* temp  = root;
        for( int i = 0;i< word.size() ; i++ )
        {
            if(temp->returnNode(word[i])==NULL)
            {
               temp->assignNode(word[i],new Node());
            }
            temp = temp->returnNode(word[i]);
        }
        temp->setEnd();
        
    }
    
    bool search(string word) {
        Node* temp  = root;
        for( int i = 0;i< word.size() ; i++ )
        {
            if(temp->returnNode(word[i])==NULL)
            {
                return false;
            }
            temp = temp->returnNode(word[i]);
        }
        
        return temp->isEnd();
    }
    
    bool startsWith(string prefix) {
        Node* temp  = root;
        for( int i = 0;i< prefix.size() ; i++ )
        {
            if(temp->returnNode(prefix[i])==NULL)
            {
                return false;
            }
            temp = temp->returnNode(prefix[i]);
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.java
class TrieNode {
    public TrieNode[] links;
    boolean isEnd;

    public TrieNode(){
        links = new TrieNode[26];
    }

    public TrieNode getNode( char c){
        return links[c-'a'];
    }
    public void addNode(char c , TrieNode temp){
        links[c-'a'] = temp;
    }
    public boolean isEnd(){
        return isEnd;
    }
    public void setEnd(){
        isEnd = true;
    }
}
class Trie {
    public TrieNode root;
    public Trie() {
        root = new TrieNode();  
    } 
    
    public void insert(String word) {
        TrieNode node = root;
        for( char c : word.toCharArray())
        {
            if( node.getNode(c)==null){
                node.addNode( c, new TrieNode());
            }
            node = node.getNode(c);
        }
        node.setEnd();
        
    }
    
    public boolean search(String word) {
         TrieNode node = root;
        for( char c : word.toCharArray())
        {
            if( node.getNode(c)==null){
                 return false;
            }
            node = node.getNode(c);
        }
         return node.isEnd()==true;
    }
    
    public boolean startsWith(String prefix) {
         TrieNode node = root;
        for( char c : prefix.toCharArray())
        {
            if( node.getNode(c)==null){
                 return false;
            }
            node = node.getNode(c);
        }

        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */